File: compiler_v2/probes/m13_custom_fns/rules/rules.go
// M13 Probe 2 — Go-authored foreign wasm module.
//
// Stand-in for what a real user would write to back a `Foreign`-aliased
// CEL custom function from Go.  The Go side of the M13 cross-language
// ABI contract (§4.5 of doc/implementation-plan/rewrite/m13-custom-fns.md).
//
// CEL declaration this targets:
//
//   bool rules.allow(this string userId, string resource);
//
// (Receiver-style with string receivers — proto messages are
// disallowed across the foreign-wasm boundary in v1; see §4.5.1
// of m13-custom-fns.md.)
//
// Synthesised overload-id (per the .celfn grammar's argkind rules):
//
//   allow_string_string
//
// Build (requires TinyGo on PATH; brew install tinygo on darwin):
//
//   tinygo build -target=wasm-unknown -no-debug \
//     -o rules.wasm ./compiler_v2/probes/m13_custom_fns/rules/
//
// The resulting `rules.wasm` is drop-in compatible with the Probe 1
// harness — swap the stub WAT for it and Probe 1's assertions still
// pass.  That's the entire point of Probe 2: prove TinyGo can hit
// the ABI without help.
//
// THIS FILE IS NOT YET BUILT BY BAZEL.  Probe 2 is gated on TinyGo
// being installed; see doc/implementation-plan/rewrite/m13-probes.md
// for the toolchain status.

package main

import "unsafe"

// CelKind tags.  Wire values frozen as of M13.A; the
// `cel.toolchain` custom section celfnc emits asserts match against
// the engine's ABI version at Engine::Instantiate.
//
// Must match compiler_v2/runtime/cel_data.h's `CelKind` enum.
const (
	celBool   = 1
	celString = 5
)

// CelValue layout — must match the 24-byte struct in
// compiler_v2/runtime/cel_data.h.  `Payload` is a raw 16-byte union;
// the layout is kind-driven and the typed bindings live in
// `celfnc`-generated stubs (Probe 4).  For Probe 2 we operate on
// the bytes directly to keep the toolchain surface minimal.
type celValue struct {
	Kind    uint32
	_pad    uint32
	Payload [16]byte
}

func slotAt(ptr uint32) *celValue {
	return (*celValue)(unsafe.Pointer(uintptr(ptr)))
}

// allow — exported wasm function.
//
// The exported name MUST match the overload-id the .celfn parser
// synthesises from the declaration.  TinyGo's `//go:wasmexport`
// directive (Go 1.21+ via TinyGo's wasm-unknown target) preserves
// the function name verbatim in the wasm `(export …)` section.
//
//go:wasmexport allow_string_string
func allow(out, userID, resource uint32) {
	u := slotAt(userID)
	r := slotAt(resource)

	// For Probe 2 the goal is "prove TinyGo can read the args and
	// write a result through the shared CelValue ABI" — not "make
	// a realistic decision."  Validate the args have the expected
	// kinds (proves we read shared memory correctly) and write
	// the bool conjunction as the result.
	//
	// A real foreign fn would decode the strings from shared memory
	// (CelSpan = {ptr, len} at payload offset 0) and inspect their
	// bytes.  The decode is straightforward — see comments inline
	// — but the probe doesn't need it to validate the ABI shape.
	ok := u.Kind == celString && r.Kind == celString

	o := slotAt(out)
	o.Kind = celBool
	// Write `1` for true, `0` for false into the leading 8 bytes of
	// the payload union — matches CelValue::payload.b (i32 at
	// offset 8) padded to 8 bytes for alignment.
	*(*uint64)(unsafe.Pointer(&o.Payload[0])) = boolU64(ok)
}

func boolU64(b bool) uint64 {
	if b {
		return 1
	}
	return 0
}

// TinyGo's wasm-unknown target still requires a main entry, but it
// does not run any startup code beyond what's strictly needed for
// the runtime — the function is effectively empty for our purposes.
func main() {}

File: compiler_v2/probes/m13_custom_fns/rules_c/rules.c
// M13 Probe 3 — Bare-C foreign wasm module.
//
// The C-no-stdlib variant of Probe 2's Go module.  Exports the same
// `allow_string_string` symbol with the same `(out_slot, *arg_slots)
// → void` ABI; the harness (m13_p3_c_test.cc) is byte-identical to
// m13_p2_test.cc except for the wasm path.
//
// Build via build_rules.sh.  Requires brew's LLVM (Apple's clang
// ships wasm32 backend but no `--target=wasm32` linker support);
// uses `/opt/homebrew/opt/llvm/bin/clang`.

#include <stdint.h>

// CelKind tags — must match compiler_v2/runtime/cel_data.h.
// Wire values frozen as of M13.A.
enum CelKind {
    CEL_BOOL = 1,
    CEL_STRING = 5,
};

// CelValue layout — 24 bytes, must match the canonical struct in
// compiler_v2/runtime/cel_data.h.  Packed for explicit layout.
typedef struct {
    uint32_t kind;
    uint32_t _pad;
    uint8_t  payload[16];
} CelValue;

// Foreign-fn entry point.  Reads CelValues at arg slot offsets,
// writes the result CelValue at out_slot.  Same shape and naming
// convention as the Go-authored counterpart in
// `compiler_v2/probes/m13_custom_fns/rules/rules.go`.
//
// CEL declaration this targets:
//
//     bool rules.allow(this string userId, string resource);
//
// Synthesised overload-id: `allow_string_string`.
__attribute__((export_name("allow_string_string")))
void allow(uint32_t out, uint32_t user_id, uint32_t resource) {
    CelValue* u = (CelValue*)(uintptr_t)user_id;
    CelValue* r = (CelValue*)(uintptr_t)resource;

    // Validate args have the expected CelKind tags (proves we read
    // shared memory correctly across the wasm boundary).  A real
    // foreign fn would decode the CelSpan {ptr, len} payload to
    // read the actual string bytes; the probe doesn't need that
    // to validate the ABI shape.
    int ok = (u->kind == CEL_STRING) && (r->kind == CEL_STRING);

    CelValue* o = (CelValue*)(uintptr_t)out;
    o->kind = CEL_BOOL;
    // Bool value lands in payload[0..8) as a 64-bit int.  Matches
    // the canonical layout (payload.b is i32 at byte offset 0,
    // padded to 8 bytes for alignment).
    uint64_t* slot = (uint64_t*)&o->payload[0];
    *slot = (uint64_t)(ok ? 1 : 0);
}
